### PR TITLE
Change bitmap inverted index to be able to be embedded in a buffer.

### PR DESCRIPTION
**Pull request by gortiz**
Current inverted implementation assumes that the index will own the channel where it is being written. At the same time the reader assumes that the index starts in the position 0 of the given buffer.

This PR changes that in order to be able to embed an inverted index in a more complex structure

### CODE REVIEW

**Comment by gortiz** [comment]
@saurabhd336 @Jackie-Jiang please review

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java [hunk 1]
```diff
@@ -83,13 +86,26 @@ public final class OffHeapBitmapInvertedIndexCreator implements DictionaryBasedI
...
+        V1Constants.Indexes.BITMAP_INVERTED_INDEX_FILE_EXTENSION);
+  }
+
+  public OffHeapBitmapInvertedIndexCreator(File indexDir, FieldSpec fieldSpec, int cardinality, int numDocs,
+      int numValues, String extension)
+      throws IOException {
+    this(indexDir, fieldSpec.getName(), fieldSpec.isSingleValueField(), cardinality, numDocs, numValues, extension);
+  }
+
+  public OffHeapBitmapInvertedIndexCreator(File indexDir, String columnName, boolean singleValue, int cardinality,
```
Comment (saurabhd336) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java:99]: Can we have a java doc for this constructor? Looks like custom file extensions are being allowed.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java [hunk 2]
```diff
@@ -83,13 +86,26 @@ public final class OffHeapBitmapInvertedIndexCreator implements DictionaryBasedI
...
+      int numDocs, int numValues, String extension)
+      throws IOException {
+    String ext = extension.equals(V1Constants.Indexes.BITMAP_INVERTED_INDEX_FILE_EXTENSION) ? "" : "." + extension;
```
Comment (saurabhd336) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java:102]: Would this is simpler if ext was Optional<String>, and 
`_invertedIndexFile

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java [hunk 3]
```diff
@@ -83,13 +86,26 @@ public final class OffHeapBitmapInvertedIndexCreator implements DictionaryBasedI
...
+    _invertedIndexFile = getDefaultFile(indexDir, columnName, extension);
+    _forwardIndexValueBufferFile = new File(indexDir, columnName + ext + FORWARD_INDEX_VALUE_BUFFER_SUFFIX);
+    _forwardIndexLengthBufferFile = new File(indexDir, columnName + ext + FORWARD_INDEX_LENGTH_BUFFER_SUFFIX);
+    _invertedIndexValueBufferFile = new File(indexDir, columnName + ext + INVERTED_INDEX_VALUE_BUFFER_SUFFIX);
+    _invertedIndexLengthBufferFile = new File(indexDir, columnName + ext + INVERTED_INDEX_LENGTH_BUFFER_SUFFIX);
```
Comment (saurabhd336) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java:107]: Can these also use getDefaultFile?

**Review by saurabhd336 — APPROVED** [review]

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/BitmapInvertedIndexWriter.java [hunk 4]
```diff
@@ -57,14 +57,33 @@ public final class BitmapInvertedIndexWriter implements Closeable {
...
+      throws IOException {
+    _ownsChannel = ownsChannel;
     int sizeForOffsets = (numBitmaps + 1) * Integer.BYTES;
     long bitmapBufferEstimate = Math.min(PESSIMISTIC_BITMAP_SIZE_ESTIMATE * numBitmaps, MAX_INITIAL_BUFFER_SIZE);
-    _fileChannel = new RandomAccessFile(outputFile, "rw").getChannel();
-    _offsetBuffer = _fileChannel.map(FileChannel.MapMode.READ_WRITE, 0, sizeForOffsets);
-    _bytesWritten = sizeForOffsets;
+    _fileChannel = fileChannel;
+    _offsetBuffer = _fileChannel.map(FileChannel.MapMode.READ_WRITE, _fileChannel.position(), sizeForOffsets);
+    _bytesWritten = sizeForOffsets + _fileChannel.position();
```
Comment (saurabhd336) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/BitmapInvertedIndexWriter.java:86]: Does this change the meaning of this variable? From _bytesWritten to _currentBufferPosition or something similar? Might warrant a change in variable name.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java (hunk 1 above)
Comment (gortiz) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java:99]: I can add javadoc, but yes, this change let you create indexes in any file without extension constraints. That is part of the idea.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/BitmapInvertedIndexWriter.java (hunk 4 above)
Comment (gortiz) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/BitmapInvertedIndexWriter.java:86]: changed

**Pull request merged**